docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,392 @@
+        extensible: sourceModel
+      data:
+        - ["global", "Member[user].Member[name]", "remote-flow"]
+
+sinkModel(type, path, kind)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adds a new taint sink. Sinks are query-specific and will usually affect one or two queries.
+
+- **type**: Name of a type from which to evaluate **path**.
+- **path**: Access path leading to the sink.
+- **kind**: Kind of sink to add. See `sink kinds <#sink-kinds>`_ for a list of supported kinds.
+
+Example:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sinkModel
+      data:
+        - ["global", "Member[eval].Argument[0]", "code-injection"]
+
+summaryModel(type, path, input, output, kind)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adds flow through a function call.
+
+- **type**: Name of a type from which to evaluate **path**.
+- **path**: Access path leading to a function call.
+- **input**: Path relative to the function call that leads to input of the flow.
+- **output**: Path relative to the function call leading to the output of the flow.
+- **kind**: Kind of summary to add. Can be **taint** for taint-propagating flow, or **value** for value-preserving flow.
+
+Example:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: summaryModel
+      data:
+        - [
+            "global",
+            "Member[decodeURIComponent]",
+            "Argument[0]",
+            "ReturnValue",
+            "taint",
+          ]
+
+typeModel(type1, type2, path)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adds a new definition of a type.
+
+- **type1**: Name of the type to define.
+- **type2**: Name of the type from which to evaluate **path**.
+- **path**: Access path leading from **type2** to **type1**.
+
+Example:
+
+.. code-block:: yaml
+
+  extensions:
+  - addsTo:
+      pack: codeql/javascript-all
+      extensible: typeModel
+    data:
+      - [
+          "mysql.Connection",
+          "@example/db",
+          "Member[getConnection].ReturnValue",
+        ]
+
+Types
+-----
+
+A type is a string that identifies a set of values.
+In each of the extension points mentioned above, the first column is always the name of a type.
+A type can be defined by adding **typeModel** tuples for that type. Additionally, the following built-in types are available:
+
+- The name of an NPM package matches imports of that package. For example, the type **express** matches the expression **require("express")**. If the package name includes dots, it must be surrounded by single quotes, such as in **'lodash.escape'**.
+- The type **global** identifies the global object, also known as **window**. In JavaScript, global variables are properties of the global object, so global variables can be identified using this type. (This type also matches imports of the NPM package named **global**, which is a package that happens to export the global object.)
```

Comment:
Is `window` a synonym for `global`?  Does it make a distinction of whether or not there is a dom environment? This seems slightly incorrect (or misleading) for non-browser environments.

Reply:
> Is `window` a synonym for `global`?

In MaD: No, you cannot write `window` as the type and expect it to match `window`.

It's just saying that the global object is also known as `window`, because in a browser environment that variable refers to the global object.

There is currently no way to specifically match references to the global object in a browser context.

Reply:
Maybe change 

> also known as **window**

to

> also known as **window** in browser contexts

Reply:
Updated to use "also known as window in browser contexts" 👍